src/components/NewCommentForm.tsx

```diff
@@ -1,8 +1,118 @@
-import React from 'react';
+import React, { FormEvent, useCallback, useState } from 'react';
+import classNames from 'classnames';
+import { addComment } from '../api/comments';
+import { CommentFormErrors, EMAIL_REGEXP, ErrorTypes } from '../constants';
+import { Comment } from '../types/Comment';
+import { Post } from '../types/Post';
+import { ErrorIcon } from './ErrorIcon';
+
+type Props = {
+  post: Post | null;
+  setErrorType: React.Dispatch<React.SetStateAction<ErrorTypes | null>>;
+  setComments: React.Dispatch<React.SetStateAction<Comment[]>>;
+};
+
+export const NewCommentForm: React.FC<Props> = ({
+  post,
+  setComments,
+  setErrorType,
+}) => {
+  const [name, setName] = useState('');
+  const [email, setEmail] = useState('');
+  const [comment, setComment] = useState('');
+  const [hasName, setHasName] = useState(true);
+  const [hasEmail, setHasEmail] = useState(true);
+  const [isEmailValid, setIsEmailValid] = useState(true);
+  const [hasComment, setHasComment] = useState(true);
+  const [isCommentAdding, setIsCommentAdding] = useState(false);
+
+  const addNewComment = useCallback(async () => {
+    setIsCommentAdding(true);
+
+    try {
+      const newComment = {
+        postId: post?.id,
+        name,
+        email,
+        body: comment,
+      };
+
+      const fetchedComment = await addComment(newComment);
+
+      setComment('');
+      setComments(prevComments => [...prevComments, fetchedComment]);
+    } catch (error) {
+      setErrorType(ErrorTypes.ADD);
+    } finally {
+      setIsCommentAdding(false);
+      setHasName(true);
+      setHasEmail(true);
+      setHasComment(true);
+    }
+  }, [name, email, comment]);
+
+  const handleFormSubmit = (event: FormEvent) => {
+    event.preventDefault();
+    let isNotValid;
+
+    if (!name) {
+      setHasName(false);
+      isNotValid = true;
+    }
+
+    if (!email) {
+      setHasEmail(false);
+      isNotValid = true;
+    }
+
+    if (!EMAIL_REGEXP(email)) {
```

Comment:
The check for email validation is incorrect. It should be '!EMAIL_REGEXP.test(email)' instead of '!EMAIL_REGEXP(email)'. The test method is used to check if a string matches a regular expression.